src/datasets/arrow_dataset.py

```diff
@@ -1723,25 +1719,36 @@ def filter(
 
         if fn_kwargs is None:
             fn_kwargs = {}
-        fn_kwargs["input_columns"] = input_columns
+        fn_kwargs.pop("input_columns", None)
 
-        # return map function
-        return self.map(
-            partial(map_function, function=function, with_indices=with_indices),
+        mask = self.map(
+            predicate_wrapper(function, with_indices, input_columns, **fn_kwargs),
             batched=True,
             with_indices=with_indices,
-            features=self.features,
             batch_size=batch_size,
-            remove_columns=remove_columns,
-            keep_in_memory=keep_in_memory,
-            load_from_cache_file=load_from_cache_file,
-            cache_file_name=cache_file_name,
-            writer_batch_size=writer_batch_size,
+            remove_columns=self.column_names,
+            keep_in_memory=compute_mask_in_memory,
             fn_kwargs=fn_kwargs,
+            cache_file_name=cache_file_name,
             num_proc=num_proc,
-            suffix_template=suffix_template,
-            new_fingerprint=new_fingerprint,
+            input_columns=input_columns,
         )
+        filtered: pa.Table = (
+            self._filter_batched(mask["mask"], filter_batch_size)
+            if filter_batch_size is not None
+            else self._data.filter(mask["mask"])
+        )
```

Comment:
This will bring all the data in memory. We'll need to use `map` to process each batch iteratively and write them on disk.

cc @SBrandeis this is the same issue as what you have for `cast`. Both `cast` and `filter` create new tables that are in-memory. Therefore we should iterate over batches and write them on disk. To make things fast we can iterate on subtables.

I think @SBrandeis is implementing a way to make it easy to transform subtables in `map`. The objective is to implement `Dataset.cast` using an API similar to this:
```python
dataset = dataset.with_format("arrow")
dataset = dataset.map(lambda table: table.cast(...), batched=True)
```

We could have the same but for `filter`.